client/consul/sync.go

```diff
@@ -0,0 +1,341 @@
+package consul
+
+import (
+	"crypto/tls"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// ConsulService allows syncing of services and checks with Consul
+type ConsulService struct {
+	client *consul.Client
+
+	task *structs.Task
+
+	services map[string]*consul.AgentService
+	checks   map[string]*structs.ServiceCheck
+
+	logger     *log.Logger
+	shutdownCh chan struct{}
+}
+
+// ConsulConfig is the configuration used to create a new ConsulService client
+type ConsulConfig struct {
+	Addr      string
+	Token     string
+	Auth      string
+	EnableSSL bool
+	VerifySSL bool
+}
+
+const (
+	// The periodic time interval for syncing services and checks with Consul
+	syncInterval = 5 * time.Second
+)
+
+// NewConsulService returns a new ConsulService
+func NewConsulService(config *ConsulConfig, logger *log.Logger) (*ConsulService, error) {
+	var err error
+	var c *consul.Client
+	cfg := consul.DefaultConfig()
+	if config.Addr != "" {
+		cfg.Address = config.Addr
+	}
+	if config.Token != "" {
+		cfg.Token = config.Token
+	}
+	if config.Auth != "" {
+		var username, password string
+		if strings.Contains(config.Auth, ":") {
+			split := strings.SplitN(config.Auth, ":", 2)
+			username = split[0]
+			password = split[1]
+		} else {
+			username = config.Auth
+		}
+
+		cfg.HttpAuth = &consul.HttpBasicAuth{
+			Username: username,
+			Password: password,
+		}
+	}
+	if config.EnableSSL {
+		cfg.Scheme = "https"
+	}
+	if config.EnableSSL && !config.VerifySSL {
+		cfg.HttpClient.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+	if c, err = consul.NewClient(cfg); err != nil {
+		return nil, err
+	}
+	consulService := ConsulService{
+		client:   c,
+		logger:   logger,
+		services: make(map[string]*consul.AgentService),
+		checks:   make(map[string]*structs.ServiceCheck),
+
+		shutdownCh: make(chan struct{}),
+	}
+	return &consulService, nil
+}
+
+// SyncTask sync the services and task with consul
+func (c *ConsulService) SyncTask(task *structs.Task) error {
+	var mErr multierror.Error
+	c.task = task
+	services := make(map[string]*consul.AgentService)
```

Comment:
Is there a way to give these different names. There are three sets of checks and services in this method:
* The ConsulService tracking
* The local tracking
* The tasks

Hard to keep track of them!